acceptance_testing.go

```diff
@@ -0,0 +1,834 @@
+
+			is.Equal(want[i], got[i])
+		}
+	})
+
+	is.True((len(gotSynchronous) > 0) != (len(gotAsynchronous) > 0)) // either Write or WriteAsync should be implemented and working (not both)
+}
+
+func (a acceptanceTest) skipIfNoSpecification(t *testing.T) {
+	if a.driver.Connector().NewSpecification == nil {
+		t.Skip("connector is missing the field NewSpecification")
+	}
+}
+
+func (a acceptanceTest) skipIfNoSource(t *testing.T) {
+	if a.driver.Connector().NewSource == nil {
+		t.Skip("connector is missing the field NewSource")
+	}
+}
+
+func (a acceptanceTest) skipIfNoDestination(t *testing.T) {
+	if a.driver.Connector().NewDestination == nil {
+		t.Skip("connector is missing the field NewDestination")
+	}
+}
+
+func (a acceptanceTest) cloneConfig(orig map[string]string) map[string]string {
+	cloned := make(map[string]string, len(orig))
+	for k, v := range orig {
+		cloned[k] = v
+	}
+	return cloned
+}
+
+func (a acceptanceTest) openSource(ctx context.Context, t *testing.T, pos Position) (source Source, cleanup func()) {
+	is := is.New(t)
+
+	source = a.driver.Connector().NewSource()
+	err := source.Configure(ctx, a.driver.SourceConfig(t))
+	is.NoErr(err)
+
+	openCtx, cancelOpenCtx := context.WithCancel(ctx)
+	err = source.Open(openCtx, pos)
+	is.NoErr(err)
+
+	// make sure connector is cleaned up only once
+	var cleanupOnce sync.Once
+	cleanup = func() {
+		cleanupOnce.Do(func() {
+			cancelOpenCtx()
+			err = source.Teardown(ctx)
+			is.NoErr(err)
+		})
+	}
+
+	return source, cleanup
+}
+
+func (a acceptanceTest) openDestination(ctx context.Context, t *testing.T) (dest Destination, cleanup func()) {
+	is := is.New(t)
+
+	dest = a.driver.Connector().NewDestination()
+	err := dest.Configure(ctx, a.driver.DestinationConfig(t))
+	is.NoErr(err)
+
+	openCtx, cancelOpenCtx := context.WithCancel(ctx)
+	err = dest.Open(openCtx)
+	is.NoErr(err)
+
+	// make sure connector is cleaned up only once
+	var cleanupOnce sync.Once
+	cleanup = func() {
+		cleanupOnce.Do(func() {
+			cancelOpenCtx()
+			err = dest.Teardown(ctx)
+			is.NoErr(err)
+		})
+	}
+
+	return dest, cleanup
+}
+
+func (a acceptanceTest) generateRecords(count int) []Record {
+	records := make([]Record, count)
+	for i := range records {
+		records[i] = a.generateRecord()
+	}
+	return records
+}
+
+func (a acceptanceTest) generateRecord() Record {
+	return Record{
+		Position:  Position(a.randString(32)), // position doesn't matter, as long as it's unique
+		Metadata:  nil,                        // TODO metadata is optional so don't enforce it to be written, still create it
+		CreatedAt: time.Unix(0, a.rand.Int63()),
+		Key:       RawData(a.randString(32)), // TODO try structured data as well
+		Payload:   RawData(a.randString(32)), // TODO try structured data as well
+	}
+}
+
+// randString generates a random string of length n.
+// (source: https://stackoverflow.com/a/31832326)
+func (a acceptanceTest) randString(n int) string {
```

Comment:
It looks like we're always calling randString with the same length parameter, so maybe we can simply use a UUID here.

Reply:
That could work, although the generated string would be somewhat limited (only `0-9`, `a-f` and `-`). I wanted to generate a string that's a bit more versatile to make sure the connector can handle it. Maybe we could go a step further and for raw data create just random bytes that are not necessarily a valid UTF-8 string (raw data can theoretically be _anything_ after all).

This reminded me of another problem related to generating data - some connectors might not support raw data at all (e.g. postgres right now works only with structured data or with JSON raw data). That might be completely valid behavior, in that case the user would have to use transforms to create structured data (e.g. use the [hoist transform](https://github.com/ConduitIO/conduit/blob/85f46e4db4041403c4bc8af9c2487c5256bc66f0/pkg/processor/transform/txfbuiltin/hoistfield.go#L49-L50)). So maybe we could give the driver the control over how records are generated and by default generate both, raw and structured data. If the connector can't handle it, the user can restrict it to either one or the other, or overwrite the method entirely and generate their own records. WDYT?

Reply:
I added a warning on top of `AcceptanceTest` since it seems pretty important to sort this out before we can actually rely on this test.

Reply:
I agree, it would be good if the driver would generate the records, due to not only types of payloads, but maybe even contents.

> Although this opens another problem - if we want to test that a destination connector knows how to write structured data, but the source only produces raw data, how do we assert we got back the expected records? sweat_smile This needs to be thought through in a separate issue.

Maybe we won't ever face such a situation, since I believe that usually a source will produce the type of payload which makes most sense for a system (S3 - raw, Pg - struct), and then the destination will require again the type of payload which makes sense for that system. 

Or maybe we need to decouple sources and destinations in the tests, so that we don't use sources to check if a destination write as OK (for example).

But yeah, I agree that this needs more thought in a separate issue. :+1: